pkg/array/array.go

```diff
@@ -124,6 +128,103 @@ func (s *Locker) UpdateArrays(configPath string, fs fs.Interface) error {
 	return nil
 }
 
+type NASCooldownTracker interface {
+	MarkFailure(nas string)
+	IsInCooldown(nas string) bool
+	ResetFailure(nas string)
+	FallbackRetry(nasList []string) string
+}
+
+type NASStatus struct {
+	Failures      int
+	CooldownUntil time.Time
+}
+
+type NASCooldown struct {
+	statusMap      map[string]*NASStatus
+	cooldownPeriod time.Duration
+	threshold      int
+	mu             sync.Mutex
+}
+
+// NewNASCooldown returns a new instance of NASCooldown.
+func NewNASCooldown(cooldownPeriod time.Duration, threshold int) *NASCooldown {
+	return &NASCooldown{
+		statusMap:      make(map[string]*NASStatus),
+		cooldownPeriod: cooldownPeriod,
+		threshold:      threshold,
+		mu:             sync.Mutex{},
+	}
+}
+
+// GetStatusMap is a getter for statusMap
+func (n *NASCooldown) GetStatusMap() map[string]*NASStatus {
+	return n.statusMap
+}
+
+// GetCooldownPeriod is a getter for cooldownPeriod
+func (n *NASCooldown) GetCooldownPeriod() time.Duration {
+	return n.cooldownPeriod
+}
+
+// GetThreshold is a getter for threshold
+func (n *NASCooldown) GetThreshold() int {
+	return n.threshold
+}
+
+// Mark NAS as failed; only enter cooldown if threshold exceeded
+func (n *NASCooldown) MarkFailure(nas string) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	status, exists := n.statusMap[nas]
+	if !exists {
+		status = &NASStatus{}
+		n.statusMap[nas] = status
+	}
+
+	status.Failures++
+	if status.Failures >= n.threshold {
+		status.CooldownUntil = time.Now().Add(n.cooldownPeriod)
+	}
+}
+
+// Check if NAS is in cooldown
+func (n *NASCooldown) IsInCooldown(nas string) bool {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	if status, exists := n.statusMap[nas]; exists {
+		return time.Now().Before(status.CooldownUntil)
+	}
+	return false
+}
+
+// Reset failure count on successful FS creation
+func (n *NASCooldown) ResetFailure(nas string) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	delete(n.statusMap, nas)
+}
+
+// Fallback logic - Retry all NAS servers, prioritizing least failed ones
+func (n *NASCooldown) FallbackRetry(nasList []string) string {
```

Comment:
do we need case FallbackRetry to handle cases where NAS entries might not exist in the statusMap.
eg sample :
func (n *NASCooldown) FallbackRetry(nasList []string) string {
    n.mu.Lock()
    defer n.mu.Unlock()

    sort.Slice(nasList, func(i, j int) bool {
        nasStatusI, existsI := n.statusMap[nasList[i]]
        nasStatusJ, existsJ := n.statusMap[nasList[j]]

        if !existsI {
            return true
        }
        if !existsJ {
            return false
        }
        return nasStatusI.Failures < nasStatusJ.Failures
    })

    return nasList[0] // Pick NAS with least failures
}